File: apps/downloadx/src/chunk.ts
import {
  QUALITY_POOR_RATIO,
  QUALITY_STALLED_RATIO,
  QUALITY_WARMUP_MS,
  UNKNOWN_SIZE_LENGTH,
} from './constants.js';
import { TypedEventEmitter } from './events.js';
import { HttpStatusError, RangeNotHonoredError, withRetry } from './retry.js';
import { SpeedTracker } from './speedTracker.js';
import type {
  ChunkLifecyclePayload,
  ChunkProgressPayload,
  ChunkQuality,
  ChunkSnapshot,
  ChunkStatus,
  DownloadConfig,
  DownloadEventMap,
  FetchResponse,
} from './types.js';

export interface ChunkParams {
  id: string;
  downloadId: string;
  url: string;
  targetFilePath: string;
  offset: number;
  length: number;
  /** Bytes already written from a previous session (resume). */
  initialDownloadedBytes: number;
  acceptsRanges: boolean;
  /** Validators sent as `If-Range` so a changed resource can't be spliced into stale bytes. */
  etag?: string | null;
  lastModified?: string | null;
  /** Live reference to download config — values read per-retry, not snapshotted. */
  global: DownloadConfig;
  /**
   * HLS segment mode. A segment chunk downloads a whole segment file from byte
   * 0 into its own `targetFilePath` (offset is always 0), is never split, and
   * — when the segment size is unknown — streams until EOF. Retry, throttle,
   * speed tracking and resume all behave exactly as for a normal chunk.
   */
  isSegment?: boolean;
  /** HLS segment: resolved source segment URI (for snapshot persistence). */
  uri?: string;
  /** HLS segment: segment duration in seconds (from #EXTINF), for ETA. */
  durationSec?: number;
  emitter: TypedEventEmitter<DownloadEventMap>;
  /** Optional throttle hook — called with bytes-just-read before write. */
  throttle?: (bytes: number, signal?: AbortSignal) => Promise<void>;
  /** Reference speed for quality classification (bytes/sec). */
  medianSpeedRef: () => number;
  /** Clock, overridable for deterministic tests. */
  now?: () => number;
}

/**
 * A single byte range being downloaded. Chunks are independent; a Download
 * owns many of them and orchestrates splits/reassignments.
 *
 * Each chunk:
 *   - issues one HTTP request per `run()` attempt
 *   - streams the body, writing to `targetFilePath` at `offset + progress`
 *   - emits progress / lifecycle / quality events
 *   - supports abort via {@link pause}, with resume honouring bytes written
 *
 * Chunks are NOT reused — a fresh instance is constructed for every attempt
 * (including after split or reassign), so state stays straightforward.
 */
export class Chunk {
  readonly id: string;
  readonly downloadId: string;
  readonly offset: number;
  /** Mutable: a chunk's length shrinks when part of it is reassigned. */
  private _length: number;

  private _status: ChunkStatus = 'pending';
  private _downloadedBytes: number;
  private _quality: ChunkQuality = 'good';
  private _retries = 0;
  private _lastError: string | undefined;
  /** Set when the failure carries scheduling meaning for the Download. */
  private _failureCode: 'range-not-honored' | null = null;

  private readonly tracker: SpeedTracker;
  private abortController: AbortController | null = null;
  private readonly params: ChunkParams;
  private readonly now: () => number;

  constructor(params: ChunkParams) {
    this.params = params;
    this.id = params.id;
    this.downloadId = params.downloadId;
    this.offset = params.offset;
    this._length = params.length;
    this._downloadedBytes = params.initialDownloadedBytes;
    this.now = params.now ?? Date.now;
    this.tracker = new SpeedTracker(params.global.speedSampleWindow, this.now);
  }

  get length(): number {
    return this._length;
  }

  get status(): ChunkStatus {
    return this._status;
  }

  get downloadedBytes(): number {
    return this._downloadedBytes;
  }

  get remainingBytes(): number {
    return Math.max(0, this._length - this._downloadedBytes);
  }

  get quality(): ChunkQuality {
    return this._quality;
  }

  get speedTracker(): SpeedTracker {
    return this.tracker;
  }

  get failureCode(): 'range-not-honored' | null {
    return this._failureCode;
  }

  /** True for HLS segment chunks — never split, written from byte 0. */
  get isSegment(): boolean {
    return this.params.isSegment ?? false;
  }

  get lastError(): string | undefined {
    return this._lastError;
  }

  snapshot(): ChunkSnapshot {
    const snap: ChunkSnapshot = {
      id: this.id,
      offset: this.offset,
      length: this._length,
      downloadedBytes: this._downloadedBytes,
      status: this._status,
      quality: this._quality,
      retries: this._retries,
    };
    if (this._lastError !== undefined) snap.lastError = this._lastError;
    // Segment chunks carry extra fields so they can be rebuilt on resume.
    if (this.params.isSegment) {
      snap.isSegment = true;
      snap.targetFilePath = this.params.targetFilePath;
      if (this.params.uri !== undefined) snap.uri = this.params.uri;
      if (this.params.durationSec !== undefined) snap.durationSec = this.params.durationSec;
    }
    return snap;
  }

  /**
   * Shrink this chunk so the tail portion can be given to another chunk.
   * Returns the byte range that was removed, or null if the chunk is too
   * close to completion to split safely.
   */
  truncateTail(minRemaining: number): { offset: number; length: number } | null {
    // Segment chunks map 1:1 to a segment file and are never split.
    if (this.isSegment) return null;
    const remaining = this.remainingBytes;
    if (remaining < minRemaining * 2) return null;
    // Cut the unclaimed half — leave at least `minRemaining` for ourselves.
    const keepFromEnd = Math.floor(remaining / 2);
    const newLength = this._length - keepFromEnd;
    const removedOffset = this.offset + newLength;
    const removedLength = keepFromEnd;
    this._length = newLength;
    return { offset: removedOffset, length: removedLength };
  }

  /** Fires abort; resume is possible if `run()` is called again afterwards. */
  pause(): void {
    if (this._status === 'completed' || this._status === 'failed') return;
    this.setStatus('paused');
    this.abortController?.abort();
  }

  /** Permanent stop — status becomes `failed` with reason. */
  fail(reason: string): void {
    this._lastError = reason;
    this.setStatus('failed');
    this.abortController?.abort();
  }

  /** Marks this chunk as reassigned (its range was moved to another chunk). */
  markReassigned(): void {
    this.setStatus('reassigned');
    this.abortController?.abort();
  }

  /**
   * Aborts the current attempt so the retry loop reissues the request from
   * the bytes already written. Used for stall recovery — the abort reason is
   * a plain Error (not AbortError), which the retry loop treats as transient.
   */
  restart(reason: string): void {
    if (this._status !== 'downloading') return;
    this.abortController?.abort(new Error(`restart: ${reason}`));
  }

  /**
   * Runs the download. Resolves when the chunk completes, fails permanently,
   * or is paused / reassigned (in which case status reflects the reason).
   */
  async run(): Promise<void> {
    if (this._status === 'completed') return;
    if (this._downloadedBytes >= this._length) {
      this.setStatus('completed');
      return;
    }

    this.setStatus('downloading');
    this.params.global.addLog({
      code: 'chunk.initialized',
      params: { id: this.id, offset: this.offset, end: this.offset + this._length },
    });

    try {
      await withRetry(
        async () => {
          await this.executeOnce();
        },
        {
          maxRetries: this.params.global.maxRetries,
          retryDelay: this.params.global.retryDelay,
          retryBackoff: this.params.global.retryBackoff,
          onRetry: (info) => {
            this._retries += 1;
            this._lastError = toMessage(info.error);
            this.params.global.addLog({
              level: 'warn',
              code: 'chunk.retry',
              params: { id: this.id, attempt: info.attempt, message: this._lastError, delayMs: info.delayMs },
            });
          },
        },
      );
      if (this._status === 'downloading') {
        this.params.global.addLog({
          code: 'chunk.completed',
          params: { id: this.id, bytes: this._downloadedBytes },
        });
        this.setStatus('completed');
      }
    } catch (err) {
      if (this._status === 'paused' || this._status === 'reassigned') return;
      if (err instanceof RangeNotHonoredError) this._failureCode = 'range-not-honored';
      this._lastError = toMessage(err);
      this.params.global.addLog({
        level: 'error',
        code: 'chunk.failed',
        params: { id: this.id, message: this._lastError },
      });
      this.setStatus('failed');
      this.params.emitter.emit('error', {
        downloadId: this.downloadId,
        chunkId: this.id,
        error: err instanceof Error ? err : new Error(this._lastError),
        fatal: false,
      });
    }
  }

  private async executeOnce(): Promise<void> {
    this.abortController = new AbortController();
    const controller = this.abortController;

    // Servers without range support always send the body from byte zero, so
    // partial progress cannot be resumed — discard it before re-requesting,
    // otherwise start-of-file bytes get written into the middle of the file.
    if (!this.params.acceptsRanges && this._downloadedBytes > 0) {
      this.params.global.addLog({
        level: 'warn',
        code: 'chunk.no-range-restart',
        params: { id: this.id, discarded: this._downloadedBytes },
      });
      this._downloadedBytes = 0;
    }

    // Idle timer: armed only while waiting on the network, so slow disks or
    // throttle waits can't trip it, and long downloads run as long as data
    // keeps arriving.
    let idleTimer: ReturnType<typeof setTimeout> | null = null;
    const clearIdle = (): void => {
      if (idleTimer !== null) {
        clearTimeout(idleTimer);
        idleTimer = null;
      }
    };
    const armIdle = (): void => {
      const ms = this.params.global.requestTimeout;
      if (ms === undefined) return;
      clearIdle();
      idleTimer = setTimeout(
        () => controller.abort(new Error(`idle timeout: no data received for ${ms}ms`)),
        ms,
      );
    };

    try {
      const rangeStart = this.offset + this._downloadedBytes;
      const headers: Record<string, string> = { ...this.params.global.headers };
      let rangeSent = false;
      const openEnded = this._length === UNKNOWN_SIZE_LENGTH;
      if (this.params.acceptsRanges) {
        headers.Range = openEnded
          ? `bytes=${rangeStart}-`
          : `bytes=${rangeStart}-${this.offset + this._length - 1}`;
        rangeSent = true;
        const validator = this.params.etag ?? this.params.lastModified;
        if (validator !== undefined && validator !== null) {
          headers['If-Range'] = validator;
        }
      }
      this.params.global.addLog({
        code: 'chunk.fetch.started',
        params: { id: this.id, url: this.params.url, range: headers.Range ?? 'none' },
      });
      armIdle();
      const res = await this.params.global.io.fetch(this.params.url, {
        method: 'GET',
        headers,
        signal: controller.signal,
      });
      clearIdle();
      if (!res.ok) {
        throw new HttpStatusError(res.status, res.statusText);
      }
      // A 200 on a ranged request means the server (or an If-Range mismatch)
      // ignored the range — consuming it would write the whole file at this
      // chunk's offset. The only safe 200 is an open-ended request from 0,
      // where the full body is byte-identical to what we asked for.
      if (rangeSent && res.status !== 206 && !(openEnded && rangeStart === 0)) {
        throw new RangeNotHonoredError();
      }
      await this.consumeBody(res, controller.signal, armIdle, clearIdle);
    } finally {
      clearIdle();
    }
  }

  private async consumeBody(
    res: FetchResponse,
    signal: AbortSignal,
    armIdle: () => void,
    clearIdle: () => void,
  ): Promise<void> {
    if (res.body === null) {
      // No stream — read whole body. Acceptable fallback for tiny chunks.
      armIdle();
      const buf = this.clampToRemaining(new Uint8Array(await res.arrayBuffer()));
      clearIdle();
      if (buf.length > 0) await this.writeBytes(buf);
      return;
    }
    const reader = res.body.getReader();
    try {
      while (true) {
        armIdle();
        const { done, value } = await reader.read();
        clearIdle();
        if (done) break;
        if (value === undefined || value.length === 0) continue;
        // Never write past our end: a split may have shrunk `_length` while
        // this stream was in flight, and a misbehaving server may send more
        // than the requested range.
        const slice = this.clampToRemaining(value);
        if (slice.length > 0) {
          if (this.params.throttle) await this.params.throttle(slice.length, signal);
          await this.writeBytes(slice);
        }
        // Stop once our (possibly shrunk) range is fully written, or when
        // status flipped during an await (pause / reassign).
        if (this._downloadedBytes >= this._length || this._status !== 'downloading') {
          try {
            await reader.cancel();
          } catch {
            /* ignore */
          }
          return;
        }
      }
    } finally {
      try {
        reader.releaseLock();
      } catch {
        /* ignore */
      }
    }
  }

  /** Slice `buf` so the write cannot exceed this chunk's current length. */
  private clampToRemaining(buf: Uint8Array): Uint8Array {
    const remaining = this._length - this._downloadedBytes;
    if (remaining <= 0) return buf.subarray(0, 0);
    return buf.length > remaining ? buf.subarray(0, remaining) : buf;
  }

  private async writeBytes(buf: Uint8Array): Promise<void> {
    const writeOffset = this.offset + this._downloadedBytes;
    await this.params.global.io.writeChunk(this.params.targetFilePath, writeOffset, buf);
    this._downloadedBytes += buf.length;
    this.tracker.record(buf.length);
    this.updateQuality();
    const progress = this.progressPayload();
    this.params.emitter.emit('chunkProgress', progress);
    this.params.emitter.emit('chunkQuality', progress);
  }

  private updateQuality(): void {
    if (!this.tracker.hasWarmedUp(QUALITY_WARMUP_MS)) {
      this._quality = 'good';
      return;
    }
    const median = this.params.medianSpeedRef();
    const mine = this.tracker.windowedSpeed;
    if (median <= 0 || mine <= 0) {
      this._quality = 'good';
      return;
    }
    const ratio = mine / median;
    if (ratio < QUALITY_STALLED_RATIO) this._quality = 'stalled';
    else if (ratio < QUALITY_POOR_RATIO) this._quality = 'poor';
    else this._quality = 'good';
  }

  private progressPayload(): ChunkProgressPayload {
    return {
      downloadId: this.downloadId,
      chunkId: this.id,
      offset: this.offset,
      length: this._length,
      downloadedBytes: this._downloadedBytes,
      instantSpeed: this.tracker.instantSpeed,
      windowedSpeed: this.tracker.windowedSpeed,
      quality: this._quality,
    };
  }

  private setStatus(next: ChunkStatus): void {
    if (this._status === next) return;
    this._status = next;
    const payload: ChunkLifecyclePayload = {
      downloadId: this.downloadId,
      chunkId: this.id,
      status: next,
    };
    this.params.emitter.emit('chunkLifecycle', payload);
  }
}

function toMessage(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

File: apps/downloadx_dart/lib/src/chunk.dart
import 'dart:async';
import 'dart:typed_data';

import 'config.dart';
import 'constants.dart';
import 'events.dart';
import 'io.dart';
import 'retry.dart';
import 'speed_tracker.dart';
import 'types.dart';

/// Immutable configuration passed to a [Chunk] at construction.
class ChunkParams {
  /// Unique identifier for this chunk.
  final String id;

  /// Identifier of the parent download.
  final String downloadId;

  /// The URL to fetch bytes from.
  final String url;

  /// Absolute path of the in-progress `.part` file.
  final String targetFilePath;

  /// Byte offset within the file where this chunk begins.
  final int offset;

  /// Number of bytes this chunk is responsible for.
  final int length;

  /// Bytes already written from a previous session (resume).
  final int initialDownloadedBytes;
  final bool acceptsRanges;

  /// Validators sent as `If-Range` so a changed resource can't be spliced.
  final String? etag;
  final String? lastModified;

  /// Live reference to download config — values read per-retry, not snapshotted.
  final DownloadConfig global;

  /// HLS segment mode. A segment chunk downloads a whole segment file from byte
  /// 0 into its own [targetFilePath] (offset is always 0), is never split, and
  /// — when the segment size is unknown — streams until EOF. Retry, throttle,
  /// speed tracking and resume all behave exactly as for a normal chunk.
  final bool isSegment;

  /// HLS segment: resolved source segment URI (for snapshot persistence).
  final String? uri;

  /// HLS segment: segment duration in seconds (from #EXTINF), for ETA.
  final double? durationSec;
  final EventEmitter emitter;

  /// Optional throttle hook — called with bytes-just-read before write.
  final Future<void> Function(int bytes, CancelToken? signal)? throttle;

  /// Reference speed for quality classification (bytes/sec).
  final double Function() medianSpeedRef;

  /// Clock, overridable for deterministic tests.
  final int Function()? now;

  /// Creates a [ChunkParams].
  const ChunkParams({
    required this.id,
    required this.downloadId,
    required this.url,
    required this.targetFilePath,
    required this.offset,
    required this.length,
    required this.initialDownloadedBytes,
    required this.acceptsRanges,
    this.etag,
    this.lastModified,
    required this.global,
    this.isSegment = false,
    this.uri,
    this.durationSec,
    required this.emitter,
    this.throttle,
    required this.medianSpeedRef,
    this.now,
  });
}

/// A single byte range being downloaded. Chunks are independent; a Download
/// owns many and orchestrates splits/reassignments.
///
/// Each chunk issues one HTTP request per `run()` attempt, streams the body
/// writing to `targetFilePath` at `offset + progress`, emits progress /
/// lifecycle / quality events, and supports abort via [pause]. Chunks are NOT
/// reused — a fresh instance is constructed for every attempt.
class Chunk {
  final String id;
  final String downloadId;
  final int offset;
  int _length;

  ChunkStatus _status = ChunkStatus.pending;
  int _downloadedBytes;
  ChunkQuality _quality = ChunkQuality.good;
  int _retries = 0;
  String? _lastError;

  /// Set when the failure carries scheduling meaning for the Download.
  String? _failureCode;

  final SpeedTracker _tracker;
  CancelToken? _abortController;
  final ChunkParams _params;
  final int Function() _now;

  Chunk(this._params)
      : id = _params.id,
        downloadId = _params.downloadId,
        offset = _params.offset,
        _length = _params.length,
        _downloadedBytes = _params.initialDownloadedBytes,
        _now = _params.now ?? _defaultNow,
        _tracker = SpeedTracker(_params.global.speedSampleWindow, _params.now);

  /// Current byte length of this chunk (may shrink when the tail is split off).
  int get length => _length;

  /// Current lifecycle state.
  ChunkStatus get status => _status;

  /// Bytes written to disk so far (including any resumed progress).
  int get downloadedBytes => _downloadedBytes;

  /// Bytes still to download, clamped to zero.
  int get remainingBytes => (_length - _downloadedBytes).clamp(0, _length);

  /// Qualitative health classification updated after each write.
  ChunkQuality get quality => _quality;

  /// Live speed tracker for this chunk.
  SpeedTracker get speedTracker => _tracker;

  /// Machine-readable failure code set on permanent failure (e.g. `'range-not-honored'`).
  String? get failureCode => _failureCode;

  /// True for HLS segment chunks — never split, written from byte 0.
  bool get isSegment => _params.isSegment;

  /// Human-readable description of the last error, or null.
  String? get lastError => _lastError;

  /// Returns an immutable snapshot of the current chunk state for persistence.
  ChunkSnapshot snapshot() => ChunkSnapshot(
        id: id,
        offset: offset,
        length: _length,
        downloadedBytes: _downloadedBytes,
        status: _status,
        quality: _quality,
        retries: _retries,
        lastError: _lastError,
        // Segment chunks carry extra fields so they can be rebuilt on resume.
        isSegment: _params.isSegment ? true : null,
        targetFilePath: _params.isSegment ? _params.targetFilePath : null,
        uri: _params.isSegment ? _params.uri : null,
        durationSec: _params.isSegment ? _params.durationSec : null,
      );

  /// Shrink this chunk so the tail portion can be given to another chunk.
  /// Returns the byte range removed, or null if too close to completion.
  ByteRange? truncateTail(int minRemaining) {
    // Segment chunks map 1:1 to a segment file and are never split.
    if (isSegment) return null;
    final remaining = remainingBytes;
    if (remaining < minRemaining * 2) return null;
    // Cut the unclaimed half — leave at least `minRemaining` for ourselves.
    final keepFromEnd = remaining ~/ 2;
    final newLength = _length - keepFromEnd;
    final removedOffset = offset + newLength;
    final removedLength = keepFromEnd;
    _length = newLength;
    return ByteRange(offset: removedOffset, length: removedLength);
  }

  /// Fires abort; resume is possible if `run()` is called again afterwards.
  void pause() {
    if (_status == ChunkStatus.completed || _status == ChunkStatus.failed) {
      return;
    }
    _setStatus(ChunkStatus.paused);
    _abortController?.cancel();
  }

  /// Permanent stop — status becomes `failed` with reason.
  void fail(String reason) {
    _lastError = reason;
    _setStatus(ChunkStatus.failed);
    _abortController?.cancel();
  }

  /// Marks this chunk as reassigned (its range was moved to another chunk).
  void markReassigned() {
    _setStatus(ChunkStatus.reassigned);
    _abortController?.cancel();
  }

  /// Aborts the current attempt so the retry loop reissues from the bytes
  /// already written. Used for stall recovery — the abort reason is a
  /// [TransientAbort] (retryable), not an [AbortError].
  void restart(String reason) {
    if (_status != ChunkStatus.downloading) return;
    _abortController?.cancel(TransientAbort('restart: $reason'));
  }

  /// Runs the download. Completes when the chunk finishes, fails permanently,
  /// or is paused / reassigned (status reflects the reason).
  Future<void> run() async {
    if (_status == ChunkStatus.completed) return;
    if (_downloadedBytes >= _length) {
      _setStatus(ChunkStatus.completed);
      return;
    }

    _setStatus(ChunkStatus.downloading);

    try {
      await withRetry<void>(
        (attempt) async {
          await _executeOnce();
        },
        RetryOptions(
          maxRetries: _params.global.maxRetries,
          retryDelay: _params.global.retryDelay,
          retryBackoff: _params.global.retryBackoff,
          onRetry: (info) {
            _retries += 1;
            _lastError = _toMessage(info.error);
            _params.emitter.emit(DiagnosticEvent(
              downloadId,
              DiagnosticPayload(
                downloadId: downloadId,
                chunkId: id,
                level: DiagnosticLevel.warn,
                code: 'chunk-retry',
                message: 'retry #$_retries in ${info.delayMs}ms: $_lastError',
                timestamp: _now(),
                data: {'attempt': info.attempt, 'delayMs': info.delayMs},
              ),
            ));
          },
        ),
      );
      if (_status == ChunkStatus.downloading) {
        _setStatus(ChunkStatus.completed);
      }
    } catch (err) {
      if (_status == ChunkStatus.paused || _status == ChunkStatus.reassigned) {
        return;
      }
      if (err is RangeNotHonoredError) _failureCode = 'range-not-honored';
      _lastError = _toMessage(err);
      _setStatus(ChunkStatus.failed);
      _params.emitter.emit(ErrorEvent(
        downloadId,
        chunkId: id,
        error: err is Exception || err is Error ? err : Exception(_lastError),
        fatal: false,
      ));
    }
  }

  Future<void> _executeOnce() async {
    final controller = CancelToken();
    _abortController = controller;

    // Servers without range support always send the body from byte zero, so
    // partial progress cannot be resumed — discard it before re-requesting.
    if (!_params.acceptsRanges && _downloadedBytes > 0) {
      _params.emitter.emit(DiagnosticEvent(
        downloadId,
        DiagnosticPayload(
          downloadId: downloadId,
          chunkId: id,
          level: DiagnosticLevel.info,
          code: 'no-range-restart',
          message:
              'server lacks range support — restarting chunk from byte 0 (discarding $_downloadedBytes bytes)',
          timestamp: _now(),
        ),
      ));
      _downloadedBytes = 0;
    }

    // Idle timer: armed only while waiting on the network, so slow disks or
    // throttle waits can't trip it, and long downloads run as long as data
    // keeps arriving.
    Timer? idleTimer;
    void clearIdle() {
      idleTimer?.cancel();
      idleTimer = null;
    }

    void armIdle() {
      final ms = _params.global.requestTimeout;
      clearIdle();
      idleTimer = Timer(Duration(milliseconds: ms), () {
        controller.cancel(
            TransientAbort('idle timeout: no data received for ${ms}ms'));
      });
    }

    try {
      final rangeStart = offset + _downloadedBytes;
      final headers = <String, String>{..._params.global.headers};
      var rangeSent = false;
      final openEnded = _length == unknownSizeLength;
      if (_params.acceptsRanges) {
        headers['Range'] = openEnded
            ? 'bytes=$rangeStart-'
            : 'bytes=$rangeStart-${offset + _length - 1}';
        rangeSent = true;
        final validator = _params.etag ?? _params.lastModified;
        if (validator != null) {
          headers['If-Range'] = validator;
        }
      }
      armIdle();
      final res = await _awaitOrAbort(
        _params.global.io.fetch(
          _params.url,
          FetchInit(method: 'GET', headers: headers, signal: controller),
        ),
        controller,
      );
      clearIdle();
      if (!res.ok) {
        throw HttpStatusError(res.status, res.statusText);
      }
      // A 200 on a ranged request means the server (or an If-Range mismatch)
      // ignored the range — consuming it would write the whole file at this
      // chunk's offset. The only safe 200 is an open-ended request from 0.
      if (rangeSent && res.status != 206 && !(openEnded && rangeStart == 0)) {
        throw RangeNotHonoredError();
      }
      await _consumeBody(res, controller, armIdle, clearIdle);
    } finally {
      clearIdle();
    }
  }

  Future<void> _consumeBody(
    FetchResponse res,
    CancelToken signal,
    void Function() armIdle,
    void Function() clearIdle,
  ) async {
    final body = res.body;
    if (body == null) {
      // No stream — read whole body. Acceptable fallback for tiny chunks.
      armIdle();
      final all = await _awaitOrAbort(res.bytes(), signal);
      clearIdle();
      final buf = _clampToRemaining(_asUint8(all));
      if (buf.isNotEmpty) await _writeBytes(buf);
      return;
    }
    final iterator = StreamIterator<List<int>>(body);
    try {
      while (true) {
        armIdle();
        final has = await _awaitOrAbort(iterator.moveNext(), signal);
        clearIdle();
        if (!has) break;
        final value = iterator.current;
        if (value.isEmpty) continue;
        // Never write past our end: a split may have shrunk `_length` while
        // this stream was in flight, and a misbehaving server may send more.
        final slice = _clampToRemaining(_asUint8(value));
        if (slice.isNotEmpty) {
          final throttle = _params.throttle;
          if (throttle != null) await throttle(slice.length, signal);
          await _writeBytes(slice);
        }
        // Stop once our (possibly shrunk) range is fully written, or when
        // status flipped during an await (pause / reassign).
        if (_downloadedBytes >= _length || _status != ChunkStatus.downloading) {
          await iterator.cancel();
          return;
        }
      }
    } finally {
      try {
        await iterator.cancel();
      } catch (_) {
        /* ignore */
      }
    }
  }

  /// Slice [buf] so the write cannot exceed this chunk's current length.
  Uint8List _clampToRemaining(Uint8List buf) {
    final remaining = _length - _downloadedBytes;
    if (remaining <= 0) return Uint8List(0);
    return buf.length > remaining
        ? Uint8List.sublistView(buf, 0, remaining)
        : buf;
  }

  Future<void> _writeBytes(Uint8List buf) async {
    final writeOffset = offset + _downloadedBytes;
    await _params.global.io
        .writeChunk(_params.targetFilePath, writeOffset, buf);
    _downloadedBytes += buf.length;
    _tracker.record(buf.length);
    _updateQuality();
    _params.emitter.emit(_chunkProgressEvent());
    _params.emitter.emit(_chunkQualityEvent());
  }

  void _updateQuality() {
    if (!_tracker.hasWarmedUp(qualityWarmupMs)) {
      _quality = ChunkQuality.good;
      return;
    }
    final median = _params.medianSpeedRef();
    final mine = _tracker.windowedSpeed;
    if (median <= 0 || mine <= 0) {
      _quality = ChunkQuality.good;
      return;
    }
    final ratio = mine / median;
    if (ratio < qualityStalledRatio) {
      _quality = ChunkQuality.stalled;
    } else if (ratio < qualityPoorRatio) {
      _quality = ChunkQuality.poor;
    } else {
      _quality = ChunkQuality.good;
    }
  }

  ChunkProgressEvent _chunkProgressEvent() => ChunkProgressEvent(
        downloadId,
        chunkId: id,
        offset: offset,
        length: _length,
        downloadedBytes: _downloadedBytes,
        instantSpeed: _tracker.instantSpeed,
        windowedSpeed: _tracker.windowedSpeed,
        quality: _quality,
      );

  ChunkQualityEvent _chunkQualityEvent() => ChunkQualityEvent(
        downloadId,
        chunkId: id,
        offset: offset,
        length: _length,
        downloadedBytes: _downloadedBytes,
        instantSpeed: _tracker.instantSpeed,
        windowedSpeed: _tracker.windowedSpeed,
        quality: _quality,
      );

  void _setStatus(ChunkStatus next) {
    if (_status == next) return;
    _status = next;
    _params.emitter
        .emit(ChunkLifecycleEvent(downloadId, chunkId: id, status: next));
  }
}

/// Completes with [future], or throws [token]'s reason if it cancels first.
Future<T> _awaitOrAbort<T>(Future<T> future, CancelToken token) {
  if (token.isCancelled) return Future<T>.error(token.reason);
  final completer = Completer<T>();
  final dispose = token.onCancel(() {
    if (!completer.isCompleted) completer.completeError(token.reason);
  });
  future.then((v) {
    dispose();
    if (!completer.isCompleted) completer.complete(v);
  }, onError: (Object e, StackTrace st) {
    dispose();
    if (!completer.isCompleted) completer.completeError(e, st);
  });
  return completer.future;
}

Uint8List _asUint8(List<int> data) =>
    data is Uint8List ? data : Uint8List.fromList(data);

String _toMessage(Object err) {
  if (err is HttpStatusError) return err.message;
  if (err is AbortError) return err.message;
  if (err is TransientAbort) return err.message;
  return err.toString();
}

int _defaultNow() => DateTime.now().millisecondsSinceEpoch;
